Add functionality to select backend

Jnelen opened the pull request:
Hi,

Here is my second try at implementing the backend selection feature. I've been quite busy, so that's why it took a bit longer than I wanted to put this pull request. Either way, below is a short summary:
I only modified graph.py, and I tried to change as little as possible.
I use environment variables to check and update backend. I also added a function to check all available backends, and to see the current backend. (That last one maybe is a bit redundant and could be removed). 
Everything seems to work fine, except there is one edge-case/"bug" with the to_graph() functionality of a molecule: if you use this before and after switching to another backend for the same molecule, it will still return the graph of the old backend because of the cache. I would propose to make use of a dictionary structure, where the (standardized) names of the backends serve as keys and the values the graphs. I could implement this quite easily, but ofcourse this would require some tweaks to molecule.py. That's why I waited with implementing this in order to first hear your opinion. Maybe there's a better way? 
Also, I didn't add any tests yes, since I'd like your feedback/opinion first! After that i'll gladly spend some more time to polish everything!

---

Close #68

RMeli commented:
> I only modified `graph.py`, and I tried to change as little as possible.

Yay! This is what I was looking for, thank you very much for taking the time to work on this. I'm really busy these days, but I'll do a proper review of this ASAP. From a quick look, it looks good!

> if you use this before and after switching to another backend for the same molecule, it will still return the graph of the old backend because of the cache.

Ah! I did not think about this issue at all, great that you bring it up. I should think about it a bit. I don't like the idea of having to change the molecule (as you guessed), but there might be no way around it. In the latter case, your proposed solution sounds sensible.

Jnelen commented:
> > I only modified `graph.py`, and I tried to change as little as possible.
> 
> Yay! This is what I was looking for, thank you very much for taking the time to work on this. I'm really busy these days, but I'll do a proper review of this ASAP. From a quick look, it looks good!
> > if you use this before and after switching to another backend for the same molecule, it will still return the graph of the old backend because of the cache.
> 
> Ah! I did not think about this issue at all, great that you bring it up. I should think about it a bit. I don't like the idea of having to change the molecule (as you guessed), but there might be no way around it. In the latter case, your proposed solution sounds sensible.

Awesome! Another point I wasn't sure of in the beginning was if the graph module had to be reimported after a backend change, but from my testing this doesn't seem to be necessary. (I thought that maybe the inital `import graph` would cache the backend dependencies). Just wanted to bring this up to be sure.

As for the code review, don't worry, there is no rush. After your review we can decide how to move forward. 
In the meantime, I will try to implement the multiprocessing function with timeout support for my own program. If it works well, I will try to port it to an (hopefully elegant) function to be integrated in the module. No promises, but i'll do my best!

RMeli commented on spyrmsd/graph.py:
```diff
@@ -1,42 +1,97 @@
+import numpy as np
+import os
+
+available_backends = []
```
```suggestion
_available_backends = []
```

I think this should be a private variable of the module, not part of the public API.

RMeli commented on spyrmsd/graph.py:
```diff
@@ -1,42 +1,97 @@
+except ImportError:
+   pass
+        
+def get_available_backends():
+    return available_backends 
```
```suggestion
def available_backends():
    return _available_backends 
```
Mainly because I have PSTD for all `set_*` and `get_*` methods that were so popular when I learned OOP in C++...

RMeli commented on spyrmsd/graph.py:
```diff
@@ -1,42 +1,97 @@
+def set_backend(backend):
+
+    ## Get the current backend
+    current_backend = str(os.environ.get("SPYRMSD_BACKEND"))
```
Do we really need to pollute the environment? Can't we get away with a local variable `_current_backend` in this module?

Jnelen replied:
You're completely right, I just updated it now!

RMeli commented on spyrmsd/graph.py:
```diff
@@ -1,42 +1,97 @@
+    ## Get the current backend
+    current_backend = str(os.environ.get("SPYRMSD_BACKEND"))
+  
+    ## Check if the backend is valid
+    if backend.lower() in ["graph_tool", "graphtool", "graph-tool", "graph tool", "gt"]:
+        backend = "graph-tool"
+    elif backend.lower() in ["networkx", "nx"]:
+        backend = "networkx"
+    else:
+        raise ValueError("This backend is not recognized or supported")
```
I think this should be its own function. I also think that it would be better to move the list of possible values in their own modules (`nx.py` and `gt.py`, respectively).

Jnelen replied:
I agree it would be cleaner to have some kind of variable which can store the aliases. However, I think in the current implementation it would give problems if we move it to the backend file (gt.py or nx.py)? Let's say graph-tool isn't installed, wouldn't trying to get the aliases from gt.py result in import errors (because gt.py imports graph-tool). Or is there an easy work around for this?

RMeli replied:
Right, I didn't think of that! In this case, maybe it's still worth putting the mapping into a (private) dictionary at the top of this module, and have a `_validate_backend` function taking `backend` as input and returning the standard backend as output? I think the check below can be also incorporated in such function  (since the backend needs to be available).

RMeli commented on spyrmsd/graph.py:
```diff
@@ -1,42 +1,97 @@
+
+    ## Check if we actually need to switch backends
+    if backend == current_backend:
+        print(f"The backend is already {backend}")
```
I don't think we want to print to standard output here. Maybe a warning would be the best here?

RMeli commented on spyrmsd/graph.py:
```diff
@@ -1,42 +1,97 @@
+        print(f"The backend is already {backend}")
+        return
+        
+    global cycle, graph_from_adjacency_matrix, lattice, match_graphs, num_edges, num_vertices, vertex_property
```
The `global` keyboard, which I never used in 10+ years of programming in Python, makes me think if there is an even cleaner solution (but possibly more complex)? 

However, if we get the API right (`set_backend`, `get_backend`, ...) we can always change how it is done later. This is why having everything very local and transparent to the user was important to me.

Jnelen replied:
I hadn't used the global keyword before either, but seems to work quite well in this case..
I'm open for any suggestions, but I can't immediately think of another/cleaner way (although as you pointed out, it might exist!)

RMeli commented on spyrmsd/graph.py:
```diff
@@ -1,42 +1,97 @@
+    ## Update the environment variable
+    os.environ["SPYRMSD_BACKEND"] = backend
+
+if len(available_backends) == 0:
+    raise ImportError("No valid backends found. Please ensure atleast Graph-Tool or NetworkX are properly installed")
+else:
+    if os.environ.get("SPYRMSD_BACKEND") is None:
+        ## Set the backend to the first available (preferred) backend         
+        set_backend(backend=available_backends[0])
```
```suggestion
if len(_available_backends) == 0:
    raise ImportError("No valid backends found. Please ensure that either graph-tool or NetworkX are installed.")
else:
    if os.environ.get("SPYRMSD_BACKEND") is None:
        ## Set the backend to the first available (preferred) backend         
        set_backend(backend=available_backends[0])
```
See comment above for the environment variable.

RMeli commented on spyrmsd/graph.py:
```diff
@@ -1,42 +1,97 @@
+        set_backend(backend=available_backends[0])
+    
+def get_backend():
+    return os.environ.get("SPYRMSD_BACKEND")
```
See comment above for the environment variable.

RMeli reviewed (changes requested):
Thanks for the PR @Jnelen, I left a few comments from a first pass of review. 

You are right, we can work first on this before adding some tests.

RMeli commented:
Mmmh, I just noticed that the test failed. Besides the style/formatting ones that should be fixable with `black`, it looks that `mypy` is not happy about this solution of re-naming functions. See [this report from CI](https://github.com/RMeli/spyrmsd/actions/runs/8102580723/job/22146163422?pr=107) for example. I'll need to look into it more closely.

Jnelen commented:
> Mmmh, I just noticed that the CI tests failed. Besides the style/formatting ones that should be fixable with `black`, it looks that `mypy` is not happy about this solution of re-naming functions. See [this report from CI](https://github.com/RMeli/spyrmsd/actions/runs/8102580723/job/22146163422?pr=107) for example. I'll need to look into it more closely.

Hi @RMeli ,

Thanks a lot for your review! It's really helpful and I'm learning a lot. I tried to implement/add your suggestions as well as I could.
I'm not really sure how to get around the mypy error, but I assume if we find another way to update/set the methods, we could also avoid having to use the `global` keyword? However, at the moment I can't really think of a better way.

Other things that still should be addressed as well I suppose are:
- adding more tests (with the `set_backend()` feature)
- fixing the molecule.to_graph() edge case caching (which gives problems  when switching backends)

Kind regards,
Jochem

RMeli reviewed (commented):
> However, at the moment I can't really think of a better way.

Me neither, and as I mentioned this is not necessarily a blocker. I'll try to look more closely at the `mypy` error, and maybe also have a look at how `matplotlib` switches between different plotting backends.

> adding more tests (with the set_backend() feature)
> fixing the molecule.to_graph() edge case caching (which gives problems when switching backends)

Definitely, but let's make sure that the design is robust before doing this. I think we are nearly there, thanks to your work!

For the molecular graph, the solution you suggested sounds sensible. Another possibility is to compute the graph on the fly, since caching it already caused bugs previously (see https://github.com/RMeli/spyrmsd/issues/17).

Jnelen commented:
> > However, at the moment I can't really think of a better way.
> 
> Me neither, and as I mentioned this is not necessarily a blocker. I'll try to look more closely at the `mypy` error, and maybe also have a look at how `matplotlib` switches between different plotting backends.
> 
> > adding more tests (with the set_backend() feature)
> > fixing the molecule.to_graph() edge case caching (which gives problems when switching backends)
> 
> Definitely, but let's make sure that the design is robust before doing this. I think we are nearly there, thanks to your work!
> 
> For the molecular graph, the solution you suggested sounds sensible. Another possibility is to compute the graph on the fly, since caching it already caused bugs previously (see #17).

Good suggestion for the _validate_backend function. I just updated to code to implement this. For the alias dictionary, I went with "generating" it, since I think it might more useful when more backends are added? If you don't like it we can also just replace it with defining the dictionary "manually". Also, i'm bad at coming up with variable names, so feel free to change them how you see fit.
I also added a warning when a backend is not installed. This seems appropriate to me, but maybe it isn't? Still it could be useful to somehow notify a user of this?

Finally regarding the caching of molecular graphs, I agree that just always doing it on the fly is a reasonable solution as well.
I'll leave the decision up to you as for what path to take!

RMeli commented on spyrmsd/graph.py:
```diff
@@ -1,42 +1,102 @@
+_networkx_aliases = ["networkx", "nx"]
+
+## Construct the alias dictionary
+_alias_backendDict = {}
```
```suggestion
_alias_to_backend = {}
```
Maybe? Not sure, but I'm not fond of mixing snake case and camel case on a single variable.

RMeli commented on spyrmsd/graph.py:
```diff
@@ -1,42 +1,102 @@
-    except ImportError:
-        raise ImportError("graph_tool or NetworkX libraries not found.")
-
-__all__ = [
-    "graph_from_adjacency_matrix",
-    "match_graphs",
-    "vertex_property",
-    "num_vertices",
-    "num_edges",
-    "lattice",
-    "cycle",
-    "adjacency_matrix_from_atomic_coordinates",
-]
```
Is it necessary to remove this, or can we still list these variables/functions as provided by the module? I think if you add it back (also with the public functions you added), `mypy` might be happy again (because it will be clear what the module provides).

RMeli commented on spyrmsd/graph.py:
```diff
@@ -1,42 +1,102 @@
+    )
+   _available_backends.append("networkx")
+except ImportError:
+   warnings.warn("The networkx backend does not seem to be installed.", stacklevel=2)
```
What is the effect of `stacklevel=2` here?

Jnelen replied:
I agree this is a weird one.. I also have this with the set_backend method.
Basically, without it, this is the warning I recieved when setting the backend looks like this:
`spyrmsd/spyrmsd/graph.py:70: UserWarning: The backend is already graph-tool
  warnings.warn(f"The backend is already {backend}")
`
So for some reason it's referring to the part where the warning is raised in the graph.py file. I found this odd, and after some googling I found the stacklevel argument. After setting the stacklevel=2, this is the warning that's produced: 
`customtest_backend.py:32: UserWarning: The backend is already graph-tool
  graph.set_backend("GT")
`
Now it refers to the part of the user's code in that actually caused the warning to happen. I found this more sensible, but I agree it's peculiar. I can remove all instances where I used stacklevel=2 and remove that part if it's more suitable.

RMeli commented on spyrmsd/graph.py:
```diff
@@ -1,42 +1,102 @@
+def _validate_backend(backend):
+    standardized_backend = _alias_backendDict.get(backend.lower())
+    if standardized_backend is None:
+        raise ValueError("This backend is not recognized or supported")
```
```suggestion
        raise ValueError(f"The {backend} backend is not recognized or supported")
```

RMeli commented on spyrmsd/graph.py:
```diff
@@ -1,42 +1,102 @@
+    return _available_backends 
+
+def set_backend(backend):
+    ## Check if the backend is valid
```
```suggestion
```
This is a bit redundant, it's already implied in the function name I think.

RMeli reviewed (changes requested):
> For the alias dictionary, I went with "generating" it, since I think it might more useful when more backends are added?

I think that's good, it's less error prone.

> If you don't like it we can also just replace it with defining 
I also added a warning when a backend is not installed.

Thanks, I think this is useful.

> I'll leave the decision up to you as for what path to take!

I'll experiment with this and see what's best.

In terms of tests, we need to add a CI configuration that has both `networkx` and `graph-tool` installed, and then a specific test (or set of tests) that only run if multiple backends are installed. I'd do something like the following (it's just a sketch):
```
def test_set_backend():
    import networks as nx
    import graph_tool as gt

    # A = ...

    set_backend("network")
    G = graph.graph_from_adjacency_matrix(A)
    assert isinstance(G, nx.Graph)

    set_backend("network")
    G = graph.graph_from_adjacency_matrix(A)
    assert isinstance(G, gt.Graph)
``` 

Can you please also run `black` and `pre-commits`, in order to fix the formatting issues that are failing in CI (`pre-commit` and `flake8` actions)? Together with the change for `mypy` I think CI should come back all green!

Thanks a lot for your work, much appreciated!

Jnelen commented:
> In terms of tests, we need to add a CI configuration that has both `networkx` and `graph-tool` installed, and then a specific test (or set of tests) that only run if multiple backends are installed. I'd do something like the following (it's just a sketch):
> 
> ```python
> def test_set_backend():
>     import networks as nx
>     import graph_tool as gt
> 
>     # A = ...
> 
>     set_backend("network")
>     G = graph.graph_from_adjacency_matrix(A)
>     assert isinstance(G, nx.Graph)
> 
>     set_backend("network")
>     G = graph.graph_from_adjacency_matrix(A)
>     assert isinstance(G, gt.Graph)
> ```
Would you want a separate set of tests? Or would you also like to modify existing tests (for example the rsmd tests) to run each of them with both backends?
> Can you please also run `black` and `pre-commits`, in order to fix the formatting issues that are failing in CI (`pre-commit` and `flake8` actions)? Together with the change for `mypy` I think CI should come back all green!
> 
Yes, I just incorporated most of the changes. I also (finally) used the pre-commit hooks. Apologies, I haven't really used this before, but i'm very glad to learn about it! After learning more about it, I think this is a very good practice which is an important part of excellent software development.

I think we're making good progress and getting close! Thank you so much for your patience and your in-depth feedback and review.

RMeli commented:
Looks like `mypy` is still unhappy. I think the path of lowest resistance at this point would be to define the functions that you import from either `gt` or `nx` to a dummy function with a nice error (just to be on the safe side), and then assign the right function to them in `set_backend`.

To be more concrete, I think something like the following should work

```python
def _dummy(*args, **kwargs):
    """
    Dummy function for backend not set.
    """
    raises NotImplementedError("No backend is set.")

num_vertices = _dummy
num_edges = _dummy
# etc...

# In set_backend, assign the function to the proper backend
# num_vertices =  nx_num_vertices
# etc...
```

> Would you want a separate set of tests? Or would you also like to modify existing tests (for example the rsmd tests) to run each of them with both backends?

I think making sure that the backend is switched and that the functions you added is the main thing in this PR. So I'd add tests where the tests for `graph.py` are. However, the only think to make sure is that they are run only if at least two backends are available. If a user installs a single backend, we still want them to be able to run the tests without failures for their own backend.

A bonus for this PR would be to print out which backend is used by default when running the tests, so that it is clear in CI.

Running all existing tests with both backends is something for another PR, which can also simplify the CI matrix quite a lot (right now, there is one configuration per graph library).

RMeli commented on spyrmsd/graph.py:
```diff
@@ -1,27 +1,97 @@
+        num_vertices = nx_num_vertices
+        vertex_property = nx_vertex_property
+
+    return backend
```
```suggestion
    # _current_backend is global
    _current_backend = backend
```

RMeli commented on spyrmsd/graph.py:
```diff
@@ -1,27 +1,97 @@
+    ## Check if we actually need to switch backends
+    if backend == _current_backend:
+        warnings.warn(f"The backend is already {backend}", stacklevel=2)
+        return _current_backend
```
```suggestion
        return
```

RMeli commented on spyrmsd/graph.py:
```diff
@@ -1,27 +1,97 @@
+    return _available_backends
+
+
+def set_backend(backend):
```
```suggestion
def set_backend(backend):
    global _current_backend
```

RMeli reviewed (changes requested):
Sorry I missed this before, but from a few tests I did, the `set_backend` function should not return anything. It should just set the `_current_backend`, which needs to be set as `global` at the top of the function (you need `global` to be able to modify such variables, while for read-only it is not needed).

RMeli commented on spyrmsd/graph.py:
```diff
@@ -1,27 +1,98 @@
+    _current_backend = backend
+
 
 __all__ = [
```
Please remove this again, I was wrong. Sorry.

Jnelen replied:
Don't worry, I wasn't sure about this either. I removed it again!

Jnelen commented:
> Looks like `mypy` is still unhappy. I think the path of lowest resistance at this point would be to define the functions that you import from either `gt` or `nx` to a dummy function with a nice error (just to be on the safe side), and then assign the right function to them in `set_backend`.
> 
> To be more concrete, I think something like the following should work
> 
> ```python
> def _dummy(*args, **kwargs):
>     """
>     Dummy function for backend not set.
>     """
>     raises NotImplementedError("No backend is set.")
> 
> num_vertices = _dummy
> num_edges = _dummy
> # etc...
> 
> # In set_backend, assign the function to the proper backend
> # num_vertices =  nx_num_vertices
> # etc...
> ```
> 
> Edit: _I'm back to the keyboard and I tested this and it indeed works. Funnily enough, when I typed `def _dummy(`, GitHub Copilot suggested a very similar function to the one suggested above._

You're right, this seems to work! Thanks a lot for the suggestion, I think it would have taken me a long time to come up with something like this myself!

> > Would you want a separate set of tests? Or would you also like to modify existing tests (for example the rsmd tests) to run each of them with both backends?
> 
> I think making sure that the backend is switched and that the functions you added work is the main thing in this PR. So I'd add tests where the tests for `graph.py` are. However, the only thing to make sure is that they are run only if at least two backends are available. If a user installs a single backend, we still want them to be able to run the tests without failures for their own backend.

I would imagine this shouldn't be too difficult by using something like `if len(graph._available_backends) > 1`. 
> A bonus for this PR would be to print out which backend is used by default when running the tests, so that it is clear in CI.
> Running all existing tests with both backends is something for another PR, which can also simplify the CI matrix quite a lot (right now, there is one configuration per graph library).

For the tests, couldn't we use the graph._available_backends, and use those as fixtures? At the start of every test (or start of the test file), we can just set the backend using the fixture as an input. I think this approach should work if you only one backend, but also would make the tests a lot more flexible when adding new backends. Maybe I'm too optimistic, but if we can make it work like this, we could reduce the CI matrix and just install both nx and gt, and it will automatically run both. But as you said, maybe it'd be cleaner to do this from a separate PR and get this to work first.

RMeli commented:
> I would imagine this shouldn't be too difficult by using something like if len(graph._available_backends) > 1.

For the simple tests of the functions you added (like the small example I did), I think using [`importskip`](https://docs.pytest.org/en/latest/how-to/skipping.html#skipping-on-a-missing-import-dependency) would be enough. 

```python
def test_set_backend() -> None:
    A = np.array([[0, 1, 1], [1, 0, 0], [1, 0, 1]])
    
    nx = pyrest.importskip("networkx")
    gt = pytest.importskip("grap_tool")

    graph.set_backend("networkx")
    assert graph.get_backend() == "networkx"
    
    Gnx = graph.graph_from_adjacency_matrix(A)
    assert isinstance(Gnx, nx.Graph)

    graph.set_backend("graph-tool")
    assert graph.get_backend() == "graph-tool"
    Ggt = graph.graph_from_adjacency_matrix(A)
    assert isinstance(Ggt, gt.Graph) 

    with pytest.raises(ValueError, match="backend is not recognized or supported"):
        graph.set_backend("unknown")
```

> For the tests, couldn't we use the graph._available_backends, and use those as fixtures? At the start of every test (or start of the test file), we can just set the backend using the fixture as an input. I think this approach should work if you only one backend, but also would make the tests a lot more flexible when adding new backends.

Yes, this is what I had in mind. Ideally, I'd like a solution that does not need to change every test function (i.e. run the fixture for setting the backend for all the tests), but the general idea is the same. However, I think this is better off on another PR to separate the concerns (if you want to work on that great, otherwise I can take care of this more boring stuff).

> I suppose that the final things to accomplish this are the updating to_graph() function and specific tests pertaining to the new backend functions?

Right, I think your suggestion of caching based on the backend is good. We can clean it up later if we then decide that building on the fly is cleaner. A test for this will be also needed.

RMeli commented on spyrmsd/graph.py:
```diff
@@ -34,9 +122,18 @@
+    )
+else:
+    if _current_backend is None:
+        ## Set the backend to the first available (preferred) backend
+        _current_backend = set_backend(backend=_available_backends[0])
```
```suggestion
        ## Set the backend to the first available (preferred) backend
        set_backend(backend=_available_backends[0])
```
After the changes I suggested, this needs to change too (otherwise `_current_backend` is initialized to `None` by default).

Jnelen replied:
good catch, thanks

Jnelen commented:
I just pushed the caching for multiple backends. I think it should be fine, but a (critical) review is always welcome. A small note is that i flush the complete cache when the `strip` function is used as this makes the most sense to me. Tomorrow I will try to work on implementing some of the specific tests related to the set_backed code as we talked about earlier.

RMeli reviewed (commented):
I received a notification with a message of yours but I can't see it here. Am I missing something?

Anyways, the PR is looking very good. I tried to add `rustworkx` as a third backend, and it was really easy to do (I'll open a PR once this is in). Thank you for all the hard work!

As you mentioned, the only thing missing is to test your functions and we are good to go!

The only additional comment I have is that while testing your PR, I realised that it might be better to have `set_beckend()` etc. available in `spyrmsd` directly instead of `spyrmsd.graph`. What do you think? If you agree it would be better, I'd make the `set_backend` a private function here, and define `set_backend` (calling `_set_backend`), `available_backends` and `get_backend` directly in `spyrmsd`.

Jnelen commented:
I had to update one of the tests to reset the cache to be an empty Dict instead of None..
All tests and mypy should pass now. I will try to work on the backend specific tests which we talked about earlier tomorrow!

Jnelen commented:
> Anyways, the PR is looking very good. I tried to add `rustworkx` as a third backend, and it was really easy to do (I'll open a PR once this is in). Thank you for all the hard work!

Oh, this is exciting! Good to see that this change is able to lead to further improvements (like new backends, hopefully improved CI, ...)
> As you mentioned, the only thing missing is to test your functions and we are good to go!
> The only additional comment I have is that while testing your PR, I realised that it might be better to have `set_beckend()` etc. available in `spyrmsd` directly instead of `spyrmsd.graph`. What do you think? If you agree it would be better, I'd make the `set_backend` a private function here, and define `set_backend` (calling `_set_backend`), `available_backends` and `get_backend` directly in `spyrmsd`.

Yes, I like this idea a lot, and it would make more sense! I just pushed the updated version. One question: do you prefer available_backends to be an attribute, or a method? either one works, but if we go with the attribute, we could even remove the available_backends function entirely. Right now I have it as an attribute (and commented out that function in graph.py), but I can very easily change it back. Also I achieved "moving" the functions by importing them in `__init__.py`. Are there better ways? Flake8 did complain (F401 '.graph._set_backend as set_backend' imported but unused), but I suppressed it for now using `# noqa: F401`. Also, isort and black fail on `__init__.py`, and when I use black to make it pass, isort fails and vice versa. So this is something I need to fix. Maybe you know a workaround/better way? Apologies for pushing it without all the tests passing, but I wanted to show the code so you could see the problem and give feedback.

Finally, I remembered that spyrmsd can also be used as a standalone tool. What if we add an optional flag for the backend, which if set simple runs the set_backend function before performing the calculation(s). Or do you think this would cause more complications, it was just a thought anyway.

RMeli commented:
Thanks for the updates @Jnelen!

> One question: do you prefer available_backends to be an attribute, or a function?

I don't mind. Attribute sounds good, but in this case let's make it public.

> Also I achieved "moving" the functions by importing them in `__init__.py`. Are there better ways?

This looks better than what I had in mind, which was just to write a wrapper function:

```python
from spyrmsd.graph import _set_backend

def set_backend(backend: str):
    _set_backend(backend)
```

Your solution seem cleaner, but if it breaks too many things the above should do.

> Apologies for pushing it without all the tests passing, but I wanted to show the code so you could see the problem and give feedback.

No need to apologise! Tests need to pass only before merging (soon!). It is much better to discuss on the actual code, as you mentioned.

> Finally, I remembered that spyrmsd can also be used as a standalone tool. What if we add an optional flag for the backend, which if set simple runs the set_backend function before performing the calculation(s). Or do you think this would cause more complications, it was just a thought anyway.

I already penciled this down on my notes as a TODO, it would definitely be a very cool feature to have! I did not mentioned it here because I did not want to over-burden you and it's something that can be added in a follow-up PR. Up to you if you want to add it here, open another follow-up PR, or leave it to me (if you are fed-up with working on this).

Jnelen commented:
> Thanks for the updates @Jnelen!
> 
> > One question: do you prefer available_backends to be an attribute, or a function?
> 
> I don't mind. Attribute sounds good, but in this case let's make it public.

Perfect, I think the current implementation should be fine then: `spyrmsd.available_backends` now provides the list with the available backends.
> > Also I achieved "moving" the functions by importing them in `__init__.py`. Are there better ways?
> 
> This looks better than what I had in mind, which was just to write a wrapper function:
> 
> ```python
> from spyrmsd.graph import _set_backend
> 
> def set_backend(backend: str):
>     _set_backend(backend)
> ```
> 
> Your solution seem cleaner, but if it breaks too many things the above should do.

I actually found that isort has a built-in profile for compatibility with black: https://pycqa.github.io/isort/docs/configuration/black_compatibility.html. I updated this in the pre-commit.yml and it passes without problems now! So if you think this is a good solution, I would keep it like this.
> > Apologies for pushing it without all the tests passing, but I wanted to show the code so you could see the problem and give feedback.
> 
> No need to apologise! Tests need to pass only before merging (soon!). It is much better to discuss on the actual code, as you mentioned.
> 
> > Finally, I remembered that spyrmsd can also be used as a standalone tool. What if we add an optional flag for the backend, which if set simple runs the set_backend function before performing the calculation(s). Or do you think this would cause more complications, it was just a thought anyway.
> 
> I already penciled this down on my notes as a TODO, it would definitely be a very cool feature to have! I did not mentioned it here because I did not want to over-burden you and it's something that can be added in a follow-up PR. Up to you if you want to add it here, open another follow-up PR (preferred), or leave it to me (if you are fed-up with working on this).

Glad to hear you think it'd be a good feature aswell. I don't think it should be too difficult to implement, but i'll leave that for a seperate PR. I will start working the backend-specific tests now as we've discussed before. I hope to have it working later!

Jnelen commented:
I just added the set_backend test. Thanks a lot for your example, it was perfect, except I needed to change the syntax (`graph.set_backend` to `spyrmsd.set_backend`). I quickly check if both backends are installed (skip the test `if len(available_backends) < 2`), however this probably isn't the optimal approach when new backends are added... But for this I suppose you'd want to update most of the CI anyway (like we talked about before with the fixtures).
Please let me know if I missed anything or other things should be added for this PR specifically!

RMeli commented:
> I actually found that isort has a built-in profile for compatibility with black

Fantastic, thanks!

> but i'll leave that for a seperate PR

Sounds good to me!

> however this probably isn't the optimal approach when new backends are added... But for this I suppose you'd want to update most of the CI anyway

Yes, exactly. This is definitely good enough to get this PR merged. The only thing that's missing is a CI configuration that actually runs the test, since in the current setup I only install one backend per configuration.

Do you want to look into this or should I? (I'm asking because some people dislike CI work and I don't want to put this burden on you.) If you would like to look into it, you'll need to
* [ ] Add a new conda configuration under [devtools/conda-env](https://github.com/RMeli/spyrmsd/tree/develop/devtools/conda-envs) that has all the bakend installed (I'd name the file `spyrmsd-test-rdkit-all.yaml`, and you can only do an RDKit configuration)
* [ ] Add this configuration to `.github/workflows/pytest.yml`
  * [ ] Add `all` (the suffix of the file suggested above) to the list of backends in [pytest.yml#L29C9-L29C10](https://github.com/RMeli/spyrmsd/blob/d498f6e0639024ca8e007d52bd1a167d12245adf/.github/workflows/pytest.yml#L29C9-L29C10)
  * [ ] Exclude the configuration for `{graphlib: "all", chemlib: "openbabel"}` in [pytest.yml#L30-L32](https://github.com/RMeli/spyrmsd/blob/d498f6e0639024ca8e007d52bd1a167d12245adf/.github/workflows/pytest.yml#L30-L32)

Jnelen commented:
> > I actually found that isort has a built-in profile for compatibility with black
> 
> Fantastic, thanks!
> 
> > but i'll leave that for a seperate PR
> 
> Sounds good to me!
> 
> > however this probably isn't the optimal approach when new backends are added... But for this I suppose you'd want to update most of the CI anyway
> 
> Yes, exactly. This is definitely good enough to get this PR merged. The only thing that's missing is a CI configuration that actually runs the test, since in the current setup I only install one backend per configuration.

Ah ofcourse, I didn't think about that, but it makes a lot of sense!
> Do you want to look into this or should I? (I'm asking because some people dislike CI work and I don't want to put this burden on you.) If you would like to look into it, you'll need to
> 
> * [ ]  Add a new conda configuration under [devtools/conda-env](https://github.com/RMeli/spyrmsd/tree/develop/devtools/conda-envs) that has all the bakend installed (I'd name the file `spyrmsd-test-rdkit-all.yaml`, and you can only do an RDKit configuration)
> * [ ]  Add this configuration to `.github/workflows/pytest.yml`
>   
>   * [ ]  Add `all` (the suffix of the file suggested above) to the list of backends in [pytest.yml#L29C9-L29C10](https://github.com/RMeli/spyrmsd/blob/d498f6e0639024ca8e007d52bd1a167d12245adf/.github/workflows/pytest.yml#L29C9-L29C10)
>   * [ ]  Exclude the configuration for `{graphlib: "all", chemlib: "openbabel"}` in [pytest.yml#L30-L32](https://github.com/RMeli/spyrmsd/blob/d498f6e0639024ca8e007d52bd1a167d12245adf/.github/workflows/pytest.yml#L30-L32)

I'll take a look! I'm planning to add CI to a project I will work on soon anyway, so I will need to learn it anyway. Why not start now haha. I think I can handle it, but If for some reason I can't get it to work I will let you know so you can take over!

RMeli commented:
> I'll take a look! I'm planning to add CI to a project I will work on soon anyway, so I will need to learn it anyway.

That's great, thanks!

RMeli commented on tests/test_graph.py:
```diff
@@ -153,3 +154,29 @@ def test_build_graph_node_features_unsupported() -> None:
+    reason="Not all of the required backends are installed",
+)
+def test_set_backend() -> None:
+    A = np.array([[0, 1, 1], [1, 0, 0], [1, 0, 1]])
+
+    import graph_tool as gt
+    import networkx as nx
```
```suggestion
    import graph_tool as gt
    import networkx as nx
    
    A = np.array([[0, 1, 1], [1, 0, 0], [1, 0, 1]])
```

RMeli commented on tests/test_molecule.py:
```diff
@@ -167,7 +167,7 @@ def test_graph_from_atomic_coordinates_perception(
 
     delattr(m, "adjacency_matrix")
-    m.G = None
+    m.G = {}
```
This is good, but we will need to add another test, similar to the one above (only if both backends are available), checking that the molecular graph is indeed re-generated for a new backend (and that at the end both are in the cache).

Jnelen replied:
Okay, I'll work on that.
Should I also test the interaction with molecule.strip and a change of backend?

RMeli replied:
Molecule strip might be too much, but I think the change of backend needs to be tested.

RMeli replied:
I'd add a very similar test to the one you added, and checking that the dictionary is populated by graphs from all backends.

RMeli commented on .github/workflows/pytest.yml:
```diff
@@ -26,10 +26,11 @@ jobs:
         exclude:
         # graph-tools does not work on Windows
         - {os: "windows-latest", graphlib: "gt"}
+        - {graphlib: "all", chemlib: "openbabel"}
```
```suggestion
        - {graphlib: "all", chemlib: "obabel"}
```

RMeli replied:
This is the reason for the latest CI failure, I think. (The configuration is not skipped.)

Jnelen replied:
> This is the reason for the latest CI failure, I think. (The configuration is not skipped.)

Yeah, I noticed it wasn't getting skipped. Dumb typo on my part!

RMeli commented on .github/workflows/pytest.yml:
```diff
@@ -26,10 +26,11 @@ jobs:
         exclude:
         # graph-tools does not work on Windows
         - {os: "windows-latest", graphlib: "gt"}
+        - {graphlib: "all", chemlib: "obabel"}
```
```suggestion
        - {graphlib: "all", chemlib: "obabel"}
        - {graphlib: "all", os: "windows-latest"}
```
CI is annoying because it's often trial-and-error. I think we need a second configuration excluding Windows when using all backends.

Jnelen replied:
I already realized the error, I just pushed it!

RMeli commented:
[Nice!](https://github.com/RMeli/spyrmsd/actions/runs/8179220936/job/22364831741?pr=107#step:6:101) Thank you very much @Jnelen for your contribution, this is very helpful!

I think we are almost ready to merge. Can you please update the `CHANGELOG` with what you did (and your GitHub handle), and add yourself to the `CITATION.ciff` file?

Jnelen commented:
> [Nice!](https://github.com/RMeli/spyrmsd/actions/runs/8179220936/job/22364831741?pr=107#step:6:101) Thank you very much @Jnelen for your contribution, this is very helpful!

It has been a great experience, so thank you for that!

> I think we are almost ready to merge. Can you please update the `CHANGELOG` with what you did (and your GitHub handle), and add yourself to the `CITATION.ciff` file?

I'm on it! Also, what about the [documentation](https://spyrmsd.readthedocs.io/en/develop/), can you update it? Or how does that work?

RMeli commented:
> It has been a great experience, so thank you for that!

Glad to hear that, thank you for bearing with me!

> Also, what about the [documentation](https://spyrmsd.readthedocs.io/en/develop/), can you update it? Or how does that work?

Dang! I always forget about the documentation. At this point I think it's better off on another PR. This one is definitely ready to go!

I'll open a couple of follow-up issues with TODOs to track these stuff. If you want to keep working on it, feel free to ping me on the new issues and I can assign them to you, otherwise I'll take care of them (eventually).

RMeli commented on CHANGELOG.md:
```diff
@@ -6,7 +6,11 @@
+
+### Added
+
+* Functionality to manually select the backend [PR  #107 | @Jnelen, @RMeli]
```
```suggestion
* Functionality to manually select the backend [PR  #107 | @Jnelen]
```
While you are at it, can you please also add a line under "changed", mentioning that the molecular graph is now cached per backend?

This is really the last thing! =)

Jnelen commented:
> > It has been a great experience, so thank you for that!
> 
> Glad to hear that, thank you for bearing with me!

I feel the opposite is more appropriate! Thank you for the patience and being very helpful in general
> > Also, what about the [documentation](https://spyrmsd.readthedocs.io/en/develop/), can you update it? Or how does that work?
> 
> Dang! I always forget about the documentation. At this point I think it's better off on another PR. This one is definitely ready to go!

Yeah I agree completely!
> I'll open a couple of follow-up issues with TODOs to track these stuff. If you want to keep working on it, feel free to ping me on the new issues and I can assign them to you, otherwise I'll take care of them (eventually).

Great! I think from here some more things can still be improved and I'm always happy to help! But the first thing I want to work on now is the parallel RMSD calculation with timeout. First I want to make it work for my application, and after that I'll see if I can adapt it to a clean spyrmsd function. After that I can try to help with some of the other stuff we talked about, like backend support for the standalone tool and cleaning up/simplifying the CI. I also have some other projects coming up however, so things might get busy for me, but I will try to keep an eye out and help where possible!

RMeli reviewed (approved):
Thanks @Jnelen, LGTM!
